{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "joyBuNA1VMDc"
      },
      "source": [
        "# **Validation of a Mitochondrial Polygenic Score (MGS) for Parkinson's Disease - iPD European Ancestry**\n",
        "\n",
        "## Project Title: Validation of a Mitochondrial Polygenic Score (MGS) for Parkinson's Disease\n",
        "\n",
        "**V:** GATK 4.3.0.0, Python 3.10.12, R 4.4.2\n",
        "\n",
        "**Note:** This notebook is for the European (EUR) ancestry group, to apply to other ancestry groups simply change the \"EUR\" to one of the following ancestries:\n",
        "\n",
        "* African Admixed (AAC)\n",
        "* African (AFR)\n",
        "* Ashkenazi Jewish (AJ)\n",
        "* American Admixed (AMR)\n",
        "* Central Asian (CAS)\n",
        "* East Asian (EAS)\n",
        "* European (EUR)\n",
        "* Middle Eastern (MDE)\n",
        "* South Asian (SAS)\n",
        "\n",
        "## Description:\n",
        "\n",
        "- [1. Getting started](#getting-started)\n",
        "- [2. Copying data to workspace](#copying-data-to-workplace)\n",
        "- [3. Formatting and QC](#formatting-and-qc)\n",
        "- [4. Calculating MGS](#calculating-mgs)\n",
        "- [5. Statistical Analyses](#statistical-analyses)\n",
        "- [6. Statistical Analyses on iPD group](#statistical-analyses-on-ipd-group)\n",
        "- [7. Saving results](#saving-results)\n",
        "\n",
        "\n",
        "For more information contact Joshua Ooi\n",
        "\n",
        "Last updated: 06/02/2025"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XRLNB17tVIQw"
      },
      "source": [
        "# Getting Started"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bH4jtSztVUyk"
      },
      "source": [
        "## Load Python Libraries"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dD_4w86lVgjK"
      },
      "outputs": [],
      "source": [
        "# Use the os package to interact with the environment (helps me find the proper paths to things)\n",
        "import os\n",
        "import sys\n",
        "\n",
        "# Bring in Pandas for Dataframe functionality (popular python packages to read in my data, manipulate, subset, filter, etc.)\n",
        "import pandas as pd\n",
        "from functools import reduce\n",
        "\n",
        "# Bring some visualization functionality (visualisation package useful when plotting stuff)\n",
        "import seaborn as sns\n",
        "\n",
        "# numpy for basics (mathematics package)\n",
        "import numpy as np\n",
        "\n",
        "# Use StringIO for working with file contents (so that it's interacting with terra cloud)\n",
        "from io import StringIO\n",
        "\n",
        "# Enable IPython to display matplotlib graphs (also a visualisation package)\n",
        "import matplotlib.pyplot as plt\n",
        "%matplotlib inline\n",
        "\n",
        "# Enable interaction with the FireCloud API (needs to be enabled in order to have terra interacting with data in the buckets)\n",
        "from firecloud import api as fapi\n",
        "\n",
        "# Import the iPython HTML rendering for displaying links to Google Cloud Console (just to display things in a jupyter notebook)\n",
        "from IPython.core.display import display, HTML\n",
        "\n",
        "# Import urllib modules for building URLs to Google Cloud Console (for interactions between terra and the cloud)\n",
        "import urllib.parse\n",
        "\n",
        "# BigQuery for querying data (for interactions between terra and the cloud)\n",
        "from google.cloud import bigquery\n",
        "\n",
        "print('Buenos Dias, Joshua!')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pgolLEe2Wzs-"
      },
      "source": [
        "## Define Python Functions to Interact with GCP/Terra"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "l-FZyyl-W0Pp"
      },
      "outputs": [],
      "source": [
        "# Utility routine for printing a shell command before executing it\n",
        "def shell_do(command):\n",
        "    print(f'Executing: {command}', file=sys.stderr)\n",
        "    !$command\n",
        "\n",
        "def shell_return(command):\n",
        "    print(f'Executing: {command}', file=sys.stderr)\n",
        "    output = !$command\n",
        "    return '\\n'.join(output)\n",
        "\n",
        "# Utility routine for printing a query before executing it\n",
        "def bq_query(query):\n",
        "    print(f'Executing: {query}', file=sys.stderr)\n",
        "    return pd.read_gbq(query, project_id=BILLING_PROJECT_ID, dialect='standard')\n",
        "\n",
        "# Utility routine for display a message and a link\n",
        "def display_html_link(description, link_text, url):\n",
        "    html = f'''\n",
        "    <p>\n",
        "    </p>\n",
        "    <p>\n",
        "    {description}\n",
        "    <a target=_blank href=\"{url}\">{link_text}</a>.\n",
        "    </p>\n",
        "    '''\n",
        "\n",
        "    display(HTML(html))\n",
        "\n",
        "# Utility routines for reading files from Google Cloud Storage\n",
        "def gcs_read_file(path):\n",
        "    \"\"\"Return the contents of a file in GCS\"\"\"\n",
        "    contents = !gsutil -u {BILLING_PROJECT_ID} cat {path}\n",
        "    return '\\n'.join(contents)\n",
        "\n",
        "def gcs_read_csv(path, sep=None):\n",
        "    \"\"\"Return a DataFrame from the contents of a delimited file in GCS\"\"\"\n",
        "    return pd.read_csv(StringIO(gcs_read_file(path)), sep=sep, engine='python')\n",
        "\n",
        "# Utility routine for displaying a message and link to Cloud Console\n",
        "def link_to_cloud_console_gcs(description, link_text, gcs_path):\n",
        "    url = '{}?{}'.format(\n",
        "        os.path.join('https://console.cloud.google.com/storage/browser',\n",
        "                     gcs_path.replace(\"gs://\",\"\")),\n",
        "        urllib.parse.urlencode({'userProject': BILLING_PROJECT_ID}))\n",
        "\n",
        "    display_html_link(description, link_text, url)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "N6UCpOulXLdz"
      },
      "source": [
        "## Initialise Work Environment Variables"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "5_fbHtEIXKCs"
      },
      "outputs": [],
      "source": [
        "# Set up billing project and data path variables\n",
        "BILLING_PROJECT_ID = os.environ['GOOGLE_PROJECT']\n",
        "WORKSPACE_NAMESPACE = os.environ['WORKSPACE_NAMESPACE']\n",
        "WORKSPACE_NAME = os.environ['WORKSPACE_NAME']\n",
        "WORKSPACE_BUCKET = os.environ['WORKSPACE_BUCKET']\n",
        "\n",
        "WORKSPACE_ATTRIBUTES = fapi.get_workspace(WORKSPACE_NAMESPACE, WORKSPACE_NAME).json().get('workspace',{}).get('attributes',{})"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "KAXkpuAvXPsH"
      },
      "source": [
        "## Accessing GP2 Data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "PZT3lotfXSrA"
      },
      "outputs": [],
      "source": [
        "##  GP2 v6.0\n",
        "GP2_RELEASE_PATH = '/GP2/release/path' ##  Enter valid GP2 Release path\n",
        "GP2_CLINICAL_RELEASE_PATH = f'{GP2_RELEASE_PATH}/clinical_data'\n",
        "GP2_META_RELEASE_PATH = f'{GP2_RELEASE_PATH}/meta_data'\n",
        "GP2_SUMSTAT_RELEASE_PATH = f'{GP2_RELEASE_PATH}/summary_statistics'\n",
        "GP2_RAW_GENO_PATH = f'{GP2_RELEASE_PATH}/raw_genotypes'\n",
        "GP2_IMPUTED_GENO_PATH = f'{GP2_RELEASE_PATH}/imputed_genotypes'\n",
        "GP2_WGS_PATH = f'{GP2_RELEASE_PATH}/wgs'\n",
        "print('GP2 v6.0')\n",
        "print(f'Path to GP2 v6.0 Clinical Data: {GP2_CLINICAL_RELEASE_PATH}')\n",
        "print(f'Path to GP2 v6.0 Raw Genotype Data: {GP2_RAW_GENO_PATH}')\n",
        "print(f'Path to GP2 v6.0 Imputed Genotype Data: {GP2_IMPUTED_GENO_PATH}')\n",
        "print(f'Path to GP2 v6.0 Metadata: {GP2_META_RELEASE_PATH}')\n",
        "print(f'Path to GP2 v6.0 WGS Data: {GP2_WGS_PATH}')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Dse97ZvLXXve"
      },
      "source": [
        "## Install PLINK"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "a34poaUfXZ5P"
      },
      "outputs": [],
      "source": [
        "%%capture\n",
        "%%bash\n",
        "# Install plink 1.9\n",
        "cd /home/jupyter/\n",
        "if test -e /home/jupyter/plink; then\n",
        "\n",
        "echo \"Plink is already installed in /home/jupyter/\"\n",
        "else\n",
        "echo \"Plink is not installed\"\n",
        "cd /home/jupyter\n",
        "\n",
        "wget http://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20190304.zip\n",
        "\n",
        "unzip -o plink_linux_x86_64_20190304.zip\n",
        "mv plink plink1.9\n",
        "\n",
        "fi"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "QcqgZBOfXbkA"
      },
      "outputs": [],
      "source": [
        "%%capture\n",
        "%%bash\n",
        "# Install plink 2.0\n",
        "cd /home/jupyter/\n",
        "if test -e /home/jupyter/plink2; then\n",
        "\n",
        "echo \"Plink2 is already installed in /home/jupyter/\"\n",
        "else\n",
        "echo \"Plink2 is not installed\"\n",
        "cd /home/jupyter/\n",
        "\n",
        "wget http://s3.amazonaws.com/plink2-assets/plink2_linux_x86_64_latest.zip\n",
        "\n",
        "unzip -o plink2_linux_x86_64_latest.zip\n",
        "\n",
        "fi"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "xKS8-M_wXc_7"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "# chmod plink 1.9 to make sure I have permission to run the program\n",
        "chmod u+x /home/jupyter/plink1.9"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "tfaG7lc0XeLN"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "# chmod plink 2.0 to make sure I have permission to run the program\n",
        "chmod u+x /home/jupyter/plink2"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VYkOmvbeXfnI"
      },
      "source": [
        "## Install R"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "0aVpFOF8XhwW"
      },
      "outputs": [],
      "source": [
        "# Install R\n",
        "! pip install --upgrade rpy2"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "R2zmEIYnXjAk"
      },
      "outputs": [],
      "source": [
        "pip install --upgrade pip"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6CPHBZ8jXkER"
      },
      "outputs": [],
      "source": [
        "%load_ext rpy2.ipython"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0saR1WDQXliB"
      },
      "source": [
        "# Copying Data to Workspace"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "aN9WZXE0XnUF"
      },
      "source": [
        "## Make a Folder"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Ic_Z54Eyjc5N"
      },
      "outputs": [],
      "source": [
        "# Assign ancestry\n",
        "ANCESTRY = \"EUR\"\n",
        "os.environ[\"ANCESTRY\"] = ANCESTRY"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "MbwgnoLMXpX2"
      },
      "outputs": [],
      "source": [
        "# Create a folder on my workspace\n",
        "print(\"Making a working directory\")\n",
        "WORK_DIR = f'/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO'\n",
        "shell_do(f'mkdir -p {WORK_DIR}')\n",
        "os.environ[\"WORK_DIR\"] = WORK_DIR"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_DNsAup1Xq5P"
      },
      "source": [
        "## Copy Files from GP2 Buckets Over to My Folder"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "QI7KdyNGXss3"
      },
      "outputs": [],
      "source": [
        "# Copy files from GP2 bucket to my workspace - Focus on the EUR\n",
        "shell_do(f'gsutil -u {BILLING_PROJECT_ID} -m cp -r {GP2_IMPUTED_GENO_PATH}/{ANCESTRY} {WORK_DIR}')\n",
        "shell_do(f'gsutil -u {BILLING_PROJECT_ID} -m cp {GP2_CLINICAL_RELEASE_PATH}/master_key_release6_final.csv {WORK_DIR}')\n",
        "shell_do(f'gsutil -u {BILLING_PROJECT_ID} -m cp {GP2_META_RELEASE_PATH}/related_samples/release6_{ANCESTRY}.related {WORK_DIR}')\n",
        "shell_do(f'gsutil -u {BILLING_PROJECT_ID} -m cp {GP2_RAW_GENO_PATH}/{ANCESTRY}/{ANCESTRY}_release6.eigenvec {WORK_DIR}')\n",
        "shell_do(f'gsutil -u {BILLING_PROJECT_ID} -m cp {GP2_RAW_GENO_PATH}/{ANCESTRY}/{ANCESTRY}_release6.eigenval {WORK_DIR}"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-5adlgT3Xu_P"
      },
      "source": [
        "## Copy non-GP2 Bucket Files Over to My Folder"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "c2vpiEIgXw3R"
      },
      "outputs": [],
      "source": [
        "# Copy files containing MGS SNPs and weights\n",
        "shell_do(f'gsutil -u {BILLING_PROJECT_ID} -m cp /enter/relevant/path/MGS_HG38.profile.txt {WORK_DIR}')\n",
        "shell_do(f'gsutil -u {BILLING_PROJECT_ID} -m cp /enter/relevant/path/MGS_HG38.profile.raw {WORK_DIR}')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "h5OpTOAGXyyD"
      },
      "source": [
        "# Formatting and QC"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pBpgubVuX1Cd"
      },
      "source": [
        "## Convert pvar files to bim"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "VinglTyTX0Yl"
      },
      "outputs": [],
      "source": [
        "%%bash -s \"$WORK_DIR\"\n",
        "cd $1\n",
        "\n",
        "for chnum in {1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22};\n",
        "do\n",
        "    /home/jupyter/plink2 \\\n",
        "    --pfile ${ANCESTRY}/chr\"$chnum\"_${ANCESTRY}_release6 \\\n",
        "    --make-bed \\\n",
        "    --out ${ANCESTRY}/MGS_chr\"$chnum\"_${ANCESTRY}_release6\n",
        "done"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SdNCFp4rX6y9"
      },
      "source": [
        "## QC *(minor allele frequency >0.01, missingness per SNP<0.05, missingness per sample <0.02 and Hardy-Weinberg equilibrium >1 × 10−50)*"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4lB2dYjjYDfN"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "\n",
        "WORK_DIR=/home/jupyter/WD_GP2_MITO_AIM1_${ANCESTRY}_PD_JO\n",
        "cd $WORK_DIR\n",
        "\n",
        "for chnum in {1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22};\n",
        "do\n",
        "    /home/jupyter/plink2 \\\n",
        "    --bfile ${ANCESTRY}/MGS_chr\"$chnum\"_${ANCESTRY}_release6 \\\n",
        "    --allow-no-sex \\\n",
        "    --maf 0.01 \\\n",
        "    --geno 0.05 \\\n",
        "    --mind 0.02 \\\n",
        "    --hwe 1e-50 \\\n",
        "    --make-bed \\\n",
        "    --out ${ANCESTRY}/MGS_chr\"$chnum\"_${ANCESTRY}_release6_QC\n",
        "done"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-etRsbz1YESP"
      },
      "source": [
        "## Concatenate the bim files to look the same as the MGS_HG38.txt file"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "DrSv6WePYG0V"
      },
      "outputs": [],
      "source": [
        "%%bash -s \"$WORK_DIR\" \"$ANCESTRY\"\n",
        "cd $1/$2\n",
        "\n",
        "for chnum in {1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22};\n",
        "do\n",
        "    awk '{print $1, $1\":\"$4, $3, $4, $5, $6}' MGS_chr\"$chnum\"_${ANCESTRY}_release6_QC.bim > temp\n",
        "    mv temp MGS_chr\"$chnum\"_${ANCESTRY}_release6_QC.bim\n",
        "done"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Mhtm1vACYO94"
      },
      "source": [
        "# Calculating MGS"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Marpp2IkYRGS"
      },
      "source": [
        "## Extract Risk SNPs"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4KDKFPILhDqg"
      },
      "outputs": [],
      "source": [
        "%%bash -s \"$WORK_DIR\"\n",
        "cd $1\n",
        "\n",
        "for chnum in {1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22};\n",
        "do\n",
        "    /home/jupyter/plink2 \\\n",
        "    --bfile ${ANCESTRY}/MGS_chr\"$chnum\"_${ANCESTRY}_release6_QC \\\n",
        "    --extract MGS_HG38.profile.txt \\\n",
        "    --make-bed \\\n",
        "    --out ${ANCESTRY}/MGS_chr\"$chnum\"_${ANCESTRY}_release6_distilled\n",
        "done"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "val-2OnEhFEE"
      },
      "source": [
        "## Merge distilled files into a single file"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "OsWhyg-uYTMJ"
      },
      "outputs": [],
      "source": [
        "%%bash -s \"$WORK_DIR\" \"$ANCESTRY\"\n",
        "cd $1/$2\n",
        "\n",
        "# Make a list that includes the files per chr that I just generated\n",
        "ls *distilled.bim > temp\n",
        "head temp"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EdnTl_NYYUwZ"
      },
      "outputs": [],
      "source": [
        "%%bash -s \"$WORK_DIR\" \"$ANCESTRY\"\n",
        "cd $1/$2\n",
        "\n",
        "# Remove *bim so i only keep the core name\n",
        "sed 's/.bim//g' temp > temp2\n",
        "cat temp2"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "uWeUeiYcYV3f"
      },
      "outputs": [],
      "source": [
        "%%bash -s \"$WORK_DIR\" \"$ANCESTRY\"\n",
        "cd $1/$2\n",
        "\n",
        "# Remove chr1 from the list so I do not merge it twice\n",
        "grep -v \"MGS_chr1_${ANCESTRY}_release6_distilled\" temp2 > files.txt\n",
        "cat files.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "b3oSsf8eYXKC"
      },
      "outputs": [],
      "source": [
        "%%bash -s \"$WORK_DIR\" \"$ANCESTRY\"\n",
        "cd $1/$2\n",
        "\n",
        "# Merge list using chr 1 as my core input\n",
        "/home/jupyter/plink1.9 \\\n",
        "--bfile MGS_chr1_${ANCESTRY}_release6_distilled \\\n",
        "--merge-list files.txt \\\n",
        "--make-bed \\\n",
        "--out MGS_ALL_${ANCESTRY}_release6"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DzpCzCoVYYka"
      },
      "source": [
        "## Calculate MGS score proper"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "T95p8SBlYaRx"
      },
      "outputs": [],
      "source": [
        "%%bash -s \"$WORK_DIR\"\n",
        "cd $1\n",
        "\n",
        "/home/jupyter/plink1.9 \\\n",
        "--bfile ${ANCESTRY}/MGS_ALL_${ANCESTRY}_release6 \\\n",
        "--score MGS_HG38.profile.txt \\\n",
        "--out ${ANCESTRY}/MGS_ALL_${ANCESTRY}_release6_score"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "a-4C7Qb3YeyE"
      },
      "source": [
        "# Statistical Analyses"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wQtI5si1YhZg"
      },
      "source": [
        "####  **Reformat Covariate Files**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "x_3YPVNpYkmu"
      },
      "outputs": [],
      "source": [
        "WORK_DIR = f'/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO'\n",
        "FULL_PATH = WORK_DIR + '/master_key_release6_final.csv'\n",
        "cov = pd.read_csv(FULL_PATH)\n",
        "cov.columns"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "AY38Yph6Ymdu"
      },
      "outputs": [],
      "source": [
        "cov_reduced = cov[['GP2sampleID', 'sex_for_qc', 'age','age_of_onset']]\n",
        "cov_reduced.rename(columns = {'GP2sampleID':'IID'}, inplace = True)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Y9EOCh-lYoMJ"
      },
      "source": [
        "####  **Merge with Eigenvec Files**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lYz0t73aYpbC"
      },
      "outputs": [],
      "source": [
        "WORK_DIR = f'/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO'\n",
        "FULL_PATH = WORK_DIR + f'/{ANCESTRY}_release6.eigenvec'\n",
        "cov2 = pd.read_csv(FULL_PATH)\n",
        "cov2.columns"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "5FLRHBMMYqzU"
      },
      "outputs": [],
      "source": [
        "WORK_DIR = f'/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO'\n",
        "FULL_PATH = WORK_DIR + f'/{ANCESTRY}_release6.eigenvec'\n",
        "cov2 = pd.read_csv(FULL_PATH)\n",
        "cov2.head()\n",
        "cov_final = cov2.merge(cov_reduced, on='IID') # Horizontal.\n",
        "cov_final.to_csv(f'/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}/gp2_covs.csv', header=True, index=False)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WDKo_LDeYtmA"
      },
      "source": [
        "## Preliminary Regression Analyses (blind to idiopathic or monogenic status)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "j4QZhmAtrw5F"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "ANCESTRY <- \"EUR\""
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "0D8bG5mXYzFi"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "# Load the glue package\n",
        "library(glue)\n",
        "ANCESTRY <- Sys.getenv(\"ANCESTRY\")\n",
        "\n",
        "work_dir <- glue(\"/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO\")\n",
        "pack <- work_dir\n",
        "\n",
        "profile <- glue(\"/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}/MGS_ALL_{ANCESTRY}_release6_score.profile\")\n",
        "covar <- glue(\"/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}/gp2_covs.csv\")\n",
        "\n",
        "temp_data <- read.table(profile, header = T)\n",
        "temp_covs <- read.csv(covar, header = T, sep=\",\")\n",
        "data <- merge(temp_data, temp_covs, by = \"IID\")\n",
        "data$CASE <- data$PHENO - 1\n",
        "data$sex_for_qc <- as.numeric(data$sex_for_qc)\n",
        "meanControls <- mean(data$SCORE[data$CASE == 0])\n",
        "sdControls <- sd(data$SCORE[data$CASE == 0])\n",
        "data$zSCORE <- (data$SCORE - meanControls)/sdControls"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "fMY-NrpUY03E"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "table(data$CASE)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "YTaHWG36Y2OX"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "data <- data[data$CASE != -10, ]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "JmTebRuaY2RL"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "table(data$CASE)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NlH52dFIY4ph"
      },
      "source": [
        "####   **Regression against disease status**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7OqSMG6ZY-Hx"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "grsTests <- glm(CASE ~ zSCORE + sex_for_qc + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + age, family=\"binomial\", data = data)\n",
        "summary(grsTests)\n",
        "\n",
        "# Extract beta and SE from the linear regression model\n",
        "beta <- coef(grsTests)[\"zSCORE\"]\n",
        "SE <- summary(grsTests)$coefficients[\"zSCORE\", \"Std. Error\"]\n",
        "\n",
        "# Calculate OR, U95, and L95\n",
        "OR <- exp(beta)\n",
        "U95 <- exp((beta) + (1.96 * SE))\n",
        "L95 <- exp((beta) - (1.96 * SE))\n",
        "\n",
        "# Print results\n",
        "print(summary(grsTests))\n",
        "\n",
        "# Print results\n",
        "print(OR)\n",
        "print(L95)\n",
        "print(U95)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XGZxC32fY_Ob"
      },
      "source": [
        "####   **Multivariable linear regression against AAO**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Zel6EJ1pZEKP"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "cases <- subset(data, CASE == 1)\n",
        "meanPop <- mean(cases$SCORE)\n",
        "sdPop <- sd(cases$SCORE)\n",
        "cases$zSCORE <- (cases$SCORE - meanPop)/sdPop\n",
        "grsTests <- lm(age_of_onset ~ zSCORE + sex_for_qc + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10, data = cases)\n",
        "summary(grsTests)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OVLZ2BtkZGpm"
      },
      "source": [
        "# Statistical Analyses on iPD Group"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "G2_-FBO7ZIR4"
      },
      "source": [
        "## Establishing the iPD group"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "---LzIRKZKYr"
      },
      "source": [
        "####   **Bring in the csv files containing the carriers of *LRRK2, PINK, PRKN* mutations**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-8dL4EObZRk0"
      },
      "outputs": [],
      "source": [
        "shell_do(f'gsutil -u {BILLING_PROJECT_ID} -m cp /enter/relevant/path/LRRK2_carriers_all.csv {WORK_DIR}')\n",
        "shell_do(f'gsutil -u {BILLING_PROJECT_ID} -m cp /enter/relevant/path/pp_hombil.csv {WORK_DIR}')\n",
        "shell_do(f'gsutil -u {BILLING_PROJECT_ID} -m cp /enter/relevant/path/pp_hombilhet.csv {WORK_DIR}')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jkF8RJ_GZVcN"
      },
      "source": [
        "####   **Start by creating a new column labelling those with pp hombil (meaning pathogenic *PRKN/PINK1* patients) as carriers**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "BYiMw5xEZhvL"
      },
      "outputs": [],
      "source": [
        "WORK_DIR = f'/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO'\n",
        "FULL_PATH = WORK_DIR + '/pp_hombil.csv'\n",
        "\n",
        "# Load the main CSV file containing subject IDs and MGS scores\n",
        "main_df = pd.read_csv(f'/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}/gp2_covs.csv')\n",
        "\n",
        "# Load the CSV file containing the list of carrier subject IDs\n",
        "carrier_df = pd.read_csv(FULL_PATH)\n",
        "\n",
        "# Create a new column in the main dataframe to indicate carrier status\n",
        "main_df[\"Carrier\"] = 0\n",
        "\n",
        "# Iterate over each subject ID in the carrier list\n",
        "for subject_id in carrier_df['IID']:\n",
        "    # Mark the corresponding rows in the main dataframe as carriers\n",
        "    main_df.loc[main_df['IID'] == subject_id, \"Carrier\"] = 1\n",
        "\n",
        "# Save the updated dataframe to a new CSV file\n",
        "main_df.to_csv(f'/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}/updated_pphombil_gp2_covs.csv', index=False)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uauQTXBgZkKZ"
      },
      "source": [
        "####   **Then add the *LRRK2* carriers to that carrier status column too**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6O9sF3nbZrE8"
      },
      "outputs": [],
      "source": [
        "# Load the updated main CSV file containing subject IDs, MGS scores, and the carrier column\n",
        "main_df = pd.read_csv(f'/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}/updated_pphombil_gp2_covs.csv')\n",
        "\n",
        "# Load the new CSV file containing the list of additional carrier subject IDs\n",
        "additional_carrier_df = pd.read_csv(f'/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/LRRK2_carriers_all.csv', delimiter='\\t')\n",
        "for subject_id in additional_carrier_df['IID']:\n",
        "    # Check if the subject ID is already in the main dataframe\n",
        "    if subject_id in main_df['IID'].values:\n",
        "        # Mark the corresponding rows as carriers without changing the designation for existing carriers\n",
        "        main_df.loc[main_df['IID'] == subject_id, \"Carrier\"] = 1\n",
        "\n",
        "# Save the updated dataframe to a new CSV file\n",
        "main_df.to_csv(f'/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}/updated_lrrk2pphombil_gp2_covs.csv', index=False)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "f3PM-bcnZtw-"
      },
      "source": [
        "####   **Some data formatting**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "FYjW7tCEZxws"
      },
      "outputs": [],
      "source": [
        "%%R -i ANCESTRY\n",
        "\n",
        "# Load the glue package\n",
        "# install.packages(\"glue\")\n",
        "library(glue)\n",
        "\n",
        "# Pass ancestry variable\n",
        "work_dir <- glue(\"/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO\")\n",
        "pack <- work_dir\n",
        "profile <- glue(\"/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}/MGS_ALL_{ANCESTRY}_release6_score.profile\")\n",
        "covar <- glue(\"/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}/updated_lrrk2pphombil_gp2_covs.csv\")\n",
        "temp_data <- read.table(profile, header = T)\n",
        "temp_covs <- read.csv(covar, header = T, sep=\",\")\n",
        "data <- merge(temp_data, temp_covs, by = \"IID\")\n",
        "data$CASE <- data$PHENO - 1\n",
        "data$sex_for_qc <- as.numeric(data$sex_for_qc)\n",
        "meanControls <- mean(data$SCORE[data$CASE == 0])\n",
        "sdControls <- sd(data$SCORE[data$CASE == 0])\n",
        "data$zSCORE <- (data$SCORE - meanControls)/sdControls"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7DraNYwFZybq"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "table(data$CASE)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "I2ULUr5PZyfE"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "data <- data[data$CASE != -10, ]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Vl7dXdtfZyit"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "table(data$CASE)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dEcqIpXjZ4Ma"
      },
      "source": [
        "## Comparing between cases and controls"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VWtDRTaYZ6-B"
      },
      "source": [
        "####   **Count means and SDs of standardised scores**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rjNs3W4rZ-pt"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "library(dplyr)\n",
        "\n",
        "# Calculate mean and SD for standardised scores\n",
        "summary_stats <- data %>%\n",
        "  group_by(CASE) %>%\n",
        "  summarize(\n",
        "    mean_standardized = mean(zSCORE, na.rm = TRUE),\n",
        "    sd_standardized = sd(zSCORE, na.rm = TRUE)\n",
        "  ) %>%\n",
        "  mutate(CASE = ifelse(CASE == 0, \"Controls\", \"Cases\")) %>%\n",
        "  rename(Group = CASE)\n",
        "\n",
        "# Print the summary statistics\n",
        "print(summary_stats)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Lc6ht-5zaAHc"
      },
      "source": [
        "####   **Deciding which statistical test to use to compare means**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TQWs3U8saELm"
      },
      "source": [
        "####   *Check for normality of control data*"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "z4AmJZclaH3j"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "library(ggplot2)\n",
        "\n",
        "# Histogram for Controls (CASE == 0)\n",
        "ggplot(data[data$CASE == 0, ], aes(x = zSCORE)) +\n",
        "  geom_histogram(binwidth = 0.2, fill = \"blue\", alpha = 0.7) +\n",
        "  ggtitle(\"Histogram of zSCORE for Controls\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "CiKqP_2PaIg8"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "library(ggplot2)\n",
        "\n",
        "# Q-Q plots to check normality for controls\n",
        "qqnorm(data$zSCORE[data$CASE == 0], main = \"Q-Q Plot for Controls\")\n",
        "qqline(data$zSCORE[data$CASE == 0])"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0lysaq9jaKmy"
      },
      "source": [
        "####   *Check for normality of cases data*"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "yLhlEAlFaOLc"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "library(ggplot2)\n",
        "\n",
        "# Histogram for Cases (CASE == 1)\n",
        "ggplot(data[data$CASE == 1, ], aes(x = zSCORE)) +\n",
        "  geom_histogram(binwidth = 0.2, fill = \"red\", alpha = 0.7) +\n",
        "  ggtitle(\"Histogram of zSCORE for Cases\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "UD0UhoxPaOsg"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "library(ggplot2)\n",
        "\n",
        "# Q-Q plots to check normality for cases\n",
        "qqnorm(data$zSCORE[data$CASE == 1], main = \"Q-Q Plot for Cases\")\n",
        "qqline(data$zSCORE[data$CASE == 1])"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "iCJXCr4_aRkH"
      },
      "source": [
        "####   *Check for equal variances*"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "eUernMbCaVFG"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "install.packages(\"car\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OKTqg5bWaV1n"
      },
      "source": [
        "####   *Check for homogeniety of variances*"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "iGRyWcYBak8x"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "data$CASE <- as.factor(data$CASE)\n",
        "\n",
        "# Run Levene's test\n",
        "library(car)\n",
        "leveneTest(zSCORE ~ CASE, data = data)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JigQTS6hamt_"
      },
      "source": [
        "## Comparisons"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zJfVkooDaoNt"
      },
      "source": [
        "####   **T-test or Mann–Whitney**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FYTAyu-kaw6m"
      },
      "source": [
        "####   *If data normally distributed and has no outliers*"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "CJ1kXsK5a4S_"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "t.test(zSCORE ~ CASE, data = data, var.equal = TRUE)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "knvu8Lfrazdn"
      },
      "source": [
        "####   *If data not normally distributed or has outliers*"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "un2uckjMa6AY"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "wilcox.test(zSCORE ~ CASE, data = data)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0PzDlMdla26M"
      },
      "source": [
        "## Visualisations"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "09DdAiy8a-fm"
      },
      "source": [
        "####   ***I quite like wesanderson colours***"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "MaNv6utVbHPP"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "install.packages(\"wesanderson\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "3nfHIW0abIVS"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "library(wesanderson)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2NHosL4gbEOR"
      },
      "source": [
        "####   ***Violin plots***"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wuBby3BxbLMo"
      },
      "outputs": [],
      "source": [
        "%%R -i ANCESTRY\n",
        "\n",
        "library(ggplot2)\n",
        "library(wesanderson)\n",
        "library(glue)\n",
        "\n",
        "colors <- wes_palette(name = \"Chevalier1\", n = 2)\n",
        "\n",
        "swapped_colors <- colors[c(2, 1)]\n",
        "\n",
        "# Create a violin plot\n",
        "p <- ggplot(data, aes(x = reorder(as.factor(CASE), zSCORE), y = zSCORE, fill = as.factor(CASE))) +\n",
        "  geom_violin(trim = FALSE) +\n",
        "  geom_boxplot(width = 0.4, fill = \"white\") +\n",
        "  theme_minimal() +\n",
        "  scale_fill_manual(values = swapped_colors) +\n",
        "  theme_bw() +\n",
        "  ylab(\"Standardized MGS\") +\n",
        "  xlab(\"\") +\n",
        "  theme(legend.position = \"none\")\n",
        "\n",
        "# Save the plot as a JPEG file\n",
        "output_jpeg <- glue(\"/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}_violin.jpeg\")\n",
        "ggsave(output_jpeg, dpi = 600, units = \"in\", height = 6, width = 6)\n",
        "\n",
        "# Display the plot\n",
        "p"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xFnmGOrdbNZQ"
      },
      "source": [
        "####   ***Density plots***"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_0BqwrX2bSDi"
      },
      "outputs": [],
      "source": [
        "%%R -i ANCESTRY\n",
        "# Density plot - probably best visualised without wesanderson colours\n",
        "library(ggplot2)\n",
        "library(glue)\n",
        "\n",
        "# Pass ancestry variable\n",
        "work_dir <- glue(\"/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO\")\n",
        "pack <- work_dir\n",
        "profile <- glue(\"/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}/MGS_ALL_{ANCESTRY}_release6_score.profile\")\n",
        "covar <- glue(\"/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}/updated_lrrk2pphombil_gp2_covs.csv\")\n",
        "temp_data <- read.table(profile, header = T)\n",
        "temp_covs <- read.csv(covar, header = T, sep=\",\")\n",
        "data <- merge(temp_data, temp_covs, by = \"IID\")\n",
        "data$CASE <- data$PHENO - 1\n",
        "data <- data[data$CASE != -10, ]\n",
        "meanControls <- mean(data$SCORE[data$CASE == 0])\n",
        "sdControls <- sd(data$SCORE[data$CASE == 0])\n",
        "head(data)\n",
        "data$zSCORE <- (data$SCORE - meanControls)/sdControls\n",
        "\n",
        "Model <- glm(CASE ~ SCORE, data = data, family = 'binomial')\n",
        "data$probDisease <- predict(Model, data, type = c(\"response\"))\n",
        "data$predicted <- ifelse(data$probDisease > 0.5, \"DISEASE\", \"CONTROL\")\n",
        "data$reported <- ifelse(data$CASE == 1, \"DISEASE\",\"CONTROL\")\n",
        "\n",
        "# Density plot\n",
        "output_png <- glue(\"/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}_density.png\")\n",
        "densPlot <- ggplot(data, aes(probDisease, fill = reported, color = reported)) + geom_density(alpha = 0.5) + theme_bw()\n",
        "ggsave(plot = densPlot, filename = output_png, width = 8, height = 5, units = \"in\", dpi = 300)\n",
        "densPlot"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2M2gXDRybX0K"
      },
      "source": [
        "## AUC analyses"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "V4WgaVDIbZTW"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "install.packages(\"pROC\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "zxgfsO-Mbalm"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "library(pROC)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "2Kw7dwlRbaoH"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "library(pROC)\n",
        "\n",
        "Model <- glm(CASE ~ SCORE, data = data, family = 'binomial')\n",
        "\n",
        "data$probDisease <- predict(Model, data, type = \"response\")\n",
        "\n",
        "roc_curve <- roc(data$CASE, data$probDisease)\n",
        "\n",
        "# print AUC value\n",
        "auc_value <- auc(roc_curve)\n",
        "cat(\"AUC Value:\", auc_value, \"\\n\")\n",
        "\n",
        "# plot ROC curve\n",
        "plot(roc_curve, main = \"ROC Curve\", col = \"blue\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "c4X6XvcPbaqu"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "# Extract the confidence interval for the AUC\n",
        "auc_ci <- ci(roc_curve)\n",
        "\n",
        "# Print AUC and CI\n",
        "cat(\"AUC:\", auc(roc_curve), \"\\n\")\n",
        "cat(\"95% CI for AUC:\", auc_ci[1], \"-\", auc_ci[3], \"\\n\")\n",
        "\n",
        "# Check if the CI includes 0.5\n",
        "if (auc_ci[1] > 0.5 | auc_ci[3] < 0.5) {\n",
        "  cat(\"The AUC is significantly different from 0.5 at the 95% confidence level.\\n\")\n",
        "} else {\n",
        "  cat(\"The AUC is not significantly different from 0.5 at the 95% confidence level.\\n\")\n",
        "}"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "JbxDMnxRbesi"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "# Extract scores for positive and negative classes\n",
        "positive_scores <- roc_curve$predictor[roc_curve$response == 1]\n",
        "negative_scores <- roc_curve$predictor[roc_curve$response == 0]\n",
        "\n",
        "# Perform a Wilcoxon rank-sum test\n",
        "wilcox_test <- wilcox.test(positive_scores, negative_scores, alternative = \"greater\")\n",
        "cat(\"P-value from Wilcoxon test:\", wilcox_test$p.value, \"\\n\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "40ldELmMbheY"
      },
      "source": [
        "# Multivariable logistic regression"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "eFx-aBtybioW"
      },
      "source": [
        "####   **Regression against disease status for those who are not carriers of mutations**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Nh_Y4N8Hbl_O"
      },
      "source": [
        "####   **Correct for age, sex, and PCs**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lXR8XQJ-bn_1"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "\n",
        "non_carrier_subset <- subset(data, Carrier == 0)\n",
        "\n",
        "# Perform logistic regression\n",
        "grsTests <- glm(CASE ~ zSCORE + sex_for_qc + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + age, family = binomial, data = non_carrier_subset)\n",
        "summary(grsTests)\n",
        "\n",
        "# Extract beta and SE from the linear regression model\n",
        "beta <- coef(grsTests)[\"zSCORE\"]\n",
        "SE <- summary(grsTests)$coefficients[\"zSCORE\", \"Std. Error\"]\n",
        "\n",
        "# Calculate OR, U95, and L95\n",
        "OR <- exp(beta)\n",
        "U95 <- exp((beta) + (1.96 * SE))\n",
        "L95 <- exp((beta) - (1.96 * SE))\n",
        "\n",
        "# Print results\n",
        "print(summary(grsTests))\n",
        "\n",
        "# Print results\n",
        "print(OR)\n",
        "print(L95)\n",
        "print(U95)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DEPiyEmPbrW8"
      },
      "source": [
        "## Multivariable linear regression"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "KN4JHPvabtxo"
      },
      "source": [
        "####   **Regression against AAO with those who are not carriers of mutations in either genes (i.e. iPD patients)**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "CEAeznWlbxUP"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "subset_cases <- subset(data, CASE == 1 & (is.na(Carrier) | Carrier == 0))\n",
        "\n",
        "# Calculate the mean and sd based on the subset\n",
        "meanPop <- mean(subset_cases$SCORE)\n",
        "sdPop <- sd(subset_cases$SCORE)\n",
        "\n",
        "# Calculate z-scores within the same subset\n",
        "subset_cases$zSCORE <- (subset_cases$SCORE - meanPop) / sdPop\n",
        "\n",
        "# Perform the linear regression on the same subset\n",
        "grsTests <- lm(age_of_onset ~ zSCORE + sex_for_qc + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10, data = subset_cases)\n",
        "summary(grsTests)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "0eSOKYttby40"
      },
      "outputs": [],
      "source": [
        "%%R\n",
        "# Assuming zSCORE and age_of_onset are in subset_cases\n",
        "\n",
        "# Calculate the correlation\n",
        "cor_test <- cor.test(subset_cases$zSCORE, subset_cases$age_of_onset)\n",
        "cor_coefficient <- cor_test$estimate\n",
        "p_value <- cor_test$p.value\n",
        "\n",
        "# Print the results\n",
        "cat(\"Correlation coefficient (r):\", cor_coefficient, \"\\n\")\n",
        "cat(\"p-value:\", p_value, \"\\n\")\n",
        "\n",
        "# Create a correlation plot\n",
        "library(ggplot2)\n",
        "\n",
        "# Generate the correlation plot\n",
        "cor_plot <- ggplot(subset_cases, aes(x = zSCORE, y = age_of_onset)) +\n",
        "  geom_point(alpha = 0.6) +\n",
        "  geom_smooth(method = \"lm\", se = FALSE, color = \"red\") +\n",
        "  theme_minimal() +\n",
        "  xlab(\"Z-Score\") +\n",
        "  ylab(\"Age of Onset\") +\n",
        "  ggtitle(paste(\"Correlation: r =\", round(cor_coefficient, 2), \", p =\", round(p_value, 3)))\n",
        "\n",
        "# Display the plot\n",
        "print(cor_plot)\n",
        "\n",
        "# Optionally save the plot\n",
        "output_jpeg <- glue(\"/home/jupyter/WD_GP2_MITO_AIM1_{ANCESTRY}_PD_JO/{ANCESTRY}_corr.jpeg\")\n",
        "ggsave(output_jpeg, plot = cor_plot, dpi = 600, units = \"in\", height = 6, width = 6)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lXVeRV-lb11Q"
      },
      "source": [
        "# Saving"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Save the final files to your workspace bucket, since we are conducting this analysis on Terra."
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
